### PR TITLE
Support File drag and drop

### PR DESCRIPTION
**Pull request by catlan**
* Generalize HTMLInputElement.files and DragEvent.dataTransfer.files access
* Register drop event and fake drag end event

Sample:

```
<div id="dropTarget" @ref="@dropTargetlement" @ondrop="@(e => OnDrop(e))" ></div>

@code {
    ElementRef  dropTargetlement;
    IFileReaderRef fileReaderReference;

    protected override void OnAfterRender()
    {
        if (fileReaderReference == null)
        {
            fileReaderReference = fileReaderService.CreateOnDropReference(dropTargetlement);
        }
    }
    
    async Task OnDrop(UIDragEventArgs e)
    {
        foreach (var file in await fileReaderReference.EnumerateFilesAsync())
        {
            // Read into memory and act
            using(MemoryStream memoryStream = await file.CreateMemoryStreamAsync(4096)) {
                // Sync calls are ok once file is in memory
                memoryStream.Read(buffer, ...)
            }
        }
    }
}

```

### CODE REVIEW

**Comment by Tewr** [comment]
Hello! Thank you for this. Conceptually I like it. I thought this API was feature complete but this looks like some nice core functionality. Might be slippery to put it in a separate assembly, so even though I'm not super familiar with the drag and drop api I'm really positive to the idea. 

I'll review it shortly but I can already say that in general it's fine. I'm just not at ease with the new method on `IFileReaderService` nor the way it is being initialized in the `FileReaderRef `constructor. 

Keep `IFileReaderService` and the  `FileReaderRef ` constructor the way it was.

Expose your method `Task InitializeAsync()` in `IFileReaderRef ` instead. 

But rename it to something better, how about  `Task RegisterDropAsync()` ?

So in your example it would be 

   ```
 protected override async Task OnAfterRenderAsync ()
    {
        if (fileReaderReference == null)
        {
            fileReaderReference = fileReaderService.CreateReference(dropTargetlement);
            await fileReaderReference.RegisterDropAsync();
        }
    }
```

File: src/Blazor.FileReader/script/FileReaderComponent.ts [hunk 1]
```diff
@@ -37,25 +37,65 @@ class FileReaderComponent {
...
+        this.dragElements.set(element, handler);
+        element.addEventListener("drop", handler);
+        return true;
+    }
 
-    public GetFileCount(element: HTMLInputElement): number {
-        if (!element.files) {
+    public GetFileCount = (element: HTMLElement): number => {
+        var files: FileList = null;
+        if (element instanceof HTMLInputElement) {
```
Comment (Tewr) [inline src/Blazor.FileReader/script/FileReaderComponent.ts:71]: This code is copy-pasted in `GetFileInfoFromElement ` and `OpenRead`, Create a new private method (`GetFiles`?)

File: src/Blazor.FileReader/FileReaderRef.cs [hunk 2]
```diff
@@ -89,10 +89,22 @@ public async Task ClearValue()
         public ElementRef ElementRef { get; private set; }
         public FileReaderJsInterop FileReaderJsInterop { get; }
 
-        internal FileReaderRef(ElementRef elementRef, FileReaderJsInterop fileReaderJsInterop)
+        private Task Initialization { get; set; }
+
+        private async Task InitializeAsync()
```
Comment (Tewr) [inline src/Blazor.FileReader/FileReaderRef.cs]: Expose this in the interface and rename it

File: src/Blazor.FileReader/FileReaderRef.cs [hunk 3]
```diff
@@ -89,10 +89,22 @@ public async Task ClearValue()
...
+            await FileReaderJsInterop.RegisterDrop(ElementRef);
+        }
+
+        internal FileReaderRef(ElementRef elementRef, FileReaderJsInterop fileReaderJsInterop, bool dropTarget = false)
         {
             this.ElementRef = elementRef;
             this.FileReaderJsInterop = fileReaderJsInterop;
+            if (dropTarget)
+            {
+                Initialization = InitializeAsync();
```
Comment (Tewr) [inline src/Blazor.FileReader/FileReaderRef.cs]: Remove from constructor

File: src/Blazor.FileReader/FileReaderService.cs [hunk 4]
```diff
@@ -40,6 +41,11 @@ public IFileReaderRef CreateReference(ElementRef inputFileElement)
             return new FileReaderRef(inputFileElement, this._fileReaderJsInterop);
         }
 
+        public IFileReaderRef CreateOnDropReference(ElementRef dropElement)
```
Comment (Tewr) [inline src/Blazor.FileReader/FileReaderService.cs]: Remove this method

File: src/Blazor.FileReader/FileReaderService.cs [hunk 5]
```diff
@@ -19,6 +19,7 @@ public interface IFileReaderService
         Task EnsureInitializedAsync();
 
         IFileReaderRef CreateReference(ElementRef inputFileElement);
+        IFileReaderRef CreateOnDropReference(ElementRef dropElement);
```
Comment (Tewr) [inline src/Blazor.FileReader/FileReaderService.cs]: Remove this method

**Review by Tewr — CHANGES_REQUESTED** [review]

**Comment by catlan** [comment]
Thank you for your feedback. I'm also glade to get rid of the initializer part.

**Review by Tewr — APPROVED** [review]

**Comment by Tewr** [comment]
I'm happy about the unregister part, the cherry on the cream. Many thanks. I'll make a new release as soon as I get the chance to test this,  probably in a few days.

I wonder if there is a way to create an E2e test with selenium for file drag and drop? Feels complicated. Time will tell...

**Pull request merged**